controllers/operatorpolicy_status.go

```diff
@@ -0,0 +1,429 @@
+	policyv1beta1 "open-cluster-management.io/config-policy-controller/api/v1beta1"
+)
+
+// updateStatus takes one condition to update, and related objects for that condition. The related
+// objects given will replace all existing relatedObjects with the same gvk. If a condition is
+// changed, the compliance will be recalculated and a compliance event will be emitted. The
+// condition and related objects can match what is already in the status - in that case, no API
+// calls are made. The `lastTransitionTime` on a condition is not considered when checking if the
+// condition has changed. If not provided, the `lastTransitionTime` will use "now". It also handles
+// preserving the `CreatedByPolicy` property on relatedObjects.
+//
+// This function assumes that all given related objects are of the same kind.
+//
+// Note that only changing the related objects will not emit a new compliance event, but will update
+// the status.
+func (r *OperatorPolicyReconciler) updateStatus(
+	ctx context.Context,
+	policy *policyv1beta1.OperatorPolicy,
+	updatedCondition metav1.Condition,
+	updatedRelatedObjs ...policyv1.RelatedObject,
+) error {
+	condChanged := false
+
+	if updatedCondition.LastTransitionTime.IsZero() {
+		updatedCondition.LastTransitionTime = metav1.Now()
+	}
+
+	condIdx, existingCondition := policy.Status.GetCondition(updatedCondition.Type)
+	if condIdx == -1 {
+		condChanged = true
+
+		// Just append, the conditions will be sorted later.
+		policy.Status.Conditions = append(policy.Status.Conditions, updatedCondition)
+	} else if conditionChanged(updatedCondition, existingCondition) {
+		condChanged = true
+
+		policy.Status.Conditions[condIdx] = updatedCondition
+	}
+
+	if condChanged {
+		updatedComplianceCondition := calculateComplianceCondition(policy)
+
+		compCondIdx, _ := policy.Status.GetCondition(updatedComplianceCondition.Type)
+		if compCondIdx == -1 {
+			policy.Status.Conditions = append(policy.Status.Conditions, updatedComplianceCondition)
+		} else {
+			policy.Status.Conditions[compCondIdx] = updatedComplianceCondition
+		}
+
+		// Sort the conditions based on their type.
+		sort.SliceStable(policy.Status.Conditions, func(i, j int) bool {
+			return policy.Status.Conditions[i].Type < policy.Status.Conditions[j].Type
+		})
+
+		if updatedComplianceCondition.Status == metav1.ConditionTrue {
+			policy.Status.ComplianceState = policyv1.Compliant
+		} else {
+			policy.Status.ComplianceState = policyv1.NonCompliant
+		}
+
+		err := r.emitComplianceEvent(ctx, policy, updatedComplianceCondition)
+		if err != nil {
+			return err
+		}
+	}
+
+	relObjsChanged := false
+
+	prevRelObjs := policy.Status.RelatedObjsOfKind(updatedRelatedObjs[0].Object.Kind)
+	if len(prevRelObjs) == len(updatedRelatedObjs) {
+		for _, prevObj := range prevRelObjs {
+			nameFound := false
+
+			for i, updatedObj := range updatedRelatedObjs {
+				if prevObj.Object.Metadata.Name == updatedObj.Object.Metadata.Name {
```

Comment:
Optional: You can make this if condition the opposite and `continue` make this less indented. The same goes for below.